### DISCUSSION TITLE
The erasure code system randomly selects BlobNodes and physical disks when creating volumes, resulting in uneven distribution of physical disks.

### DISCUSSION THREAD

**Discussion by FourSpaces** [cubefs/cubefs#3542] [Q&A]
When there are few nodes and disks, the uncontrollability of random numbers may lead to uneven distribution of volumes and disks, and multiple volumes of disks may contain the same disk. The read and write capabilities of the cluster are affected by several disk bottlenecks.

Why not use a fixed disk grouping method to organize disks and nodes, and use polling or hashing to obtain a disk allocation list from the disk group for volume allocation. In this way, the distribution of volumes can be controlled and all disks in the cluster can be fully utilized.

**Comment by tangdeyi** [comment]
In fact, the main problem is that when the cluster is small, the distribution of volumes on the disk cannot truly reflect the balance of the allocation algorithm. Imagine such a scenario. As the cluster water level rises, it is inevitable that the storage nodes will need to be expanded. At this time, if you want to fully utilize the disk IO capacity of the cluster, you must perform a balancing task. However, it is also possible that the current cluster IO capacity has been met, so it can be unbalanced (because the balancing task will also occupy internal IO). At this time, the advantage of the weighted allocation algorithm is reflected. The newly created volumes will be dispersed to the newly expanded nodes first, and the polling or hashing method you mentioned does not have such capabilities.

**Reply by FourSpaces** [reply]
@tangdeyi  When the cluster is small, the random allocation method is indeed very uneven. The disk grouping polling method I proposed does not conflict with node load allocation. In the stage of filtering disk groups, disk groups that do not meet the load will be screened out, and newly added nodes (disk groups) and old nodes (disk groups) that meet the load conditions will be left behind. Volumes will be allocated from the filtered disk groups using polling or hash remainder.

**Reply by tangdeyi** [reply]
There is a problem with the polling or hashing algorithm you proposed. As the cluster water level rises, the number of volume units on the disk cannot gradually become balanced. Instead, some disks have no new volume units to allocate, but some disks still have many available volume units. The weight algorithm based on available volume units can solve this problem.

**Comment by FourSpaces** [comment]
ok

**Comment by FourSpaces** [comment]
@tangdeyi  I see in the code that disk allocation is not based on free blocks, but on random values, which is inconsistent with the previous statement. I believe that sorting can be done based on disk free blocks, which allows the current node to allocate disks multiple times without duplication, and also enables load balancing

https://github.com/cubefs/cubefs/blob/b8815f9423a8eb6d11f2227a8bc1e318462cfc27/blobstore/clustermgr/diskmgr/alloc.go#L298C78-L298C79


I think the following writing is necessary to achieve the function of loading idle blocks on the same node

```
func (d *blobNodeAllocator) allocDisk(ctx context.Context, excludes map[proto.DiskID]*diskItem) (chosenDisk *diskItem) {
	span := trace.SpanFromContextSafe(ctx)
	totalFreeChunk := atomic.LoadInt64(&d.freeChunk)
	if totalFreeChunk <= 0 {
		return nil
	}
	total := len(d.disks)
	randTotal := total
	disks := make([]*diskItem, 0, total)
	disks = append(disks, d.disks...)
        // Sort the free blocks on the disk, with the ones with more free blocks ranking first
	sort.Slice(disks, func(i, j int) bool {
                 disks[i].lock.RLock()
		disks[j].lock.RLock()
		defer disks[i].lock.RUnlock()
		defer disks[j].lock.RUnlock()

		if disks[i].info.FreeChunkCnt == disks[j].info.FreeChunkCnt {
			return disks[i].info.DiskID > disks[j].info.DiskID
		} else {
			return disks[i].info.FreeChunkCnt > disks[j].info.FreeChunkCnt
		}
	})
	
	for i := 0; i < total; i++ {
		chosenDisk = func() *diskItem {
			//randNum := rand.Intn(randTotal)
			randNum := i
			defer func() {
				disks[randTotal-1], disks[randNum] = disks[randNum], disks[randTotal-1]
				randTotal--
			}()
			disk := disks[randNum]
			disk.lock.RLock()
			defer disk.lock.RUnlock()
			freeChunk := disk.info.FreeChunkCnt
			if freeChunk <= 0 {
				return nil
			}
			// ignore not writable disk
			if !disk.isWritable() {
				span.Debugf("disk %d is not writable, is it expired: %v", disk.diskID, disk.isExpire())
				return nil
			}

			if _, ok := excludes[disk.diskID]; !ok {
				span.Debugf("chosen disk: %#v", disk.info)
				return disk
			}
			return nil
		}()
		if chosenDisk != nil {
			return
		}
	}
	return chosenDisk
}
```

**Reply by tangdeyi** [reply]
The selection of disks within a node is as you said, because the storage node will be selected before the disk is selected, and the logic of selecting the node is based on the free chunk of the node. In this way, even if the disk is selected randomly within the node, there will be no data skew.

https://github.com/cubefs/cubefs/blob/b8815f9423a8eb6d11f2227a8bc1e318462cfc27/blobstore/clustermgr/diskmgr/alloc.go#L490
https://github.com/cubefs/cubefs/blob/b8815f9423a8eb6d11f2227a8bc1e318462cfc27/blobstore/clustermgr/diskmgr/alloc.go#L386

**Reply by FourSpaces** [reply]
I think using free blocks as a selection criterion when selecting disks within a node can achieve a more balanced utilization of disks, which may result in better performance
